Restructure classes for table database objects

trueqbit opened the pull request:
Sharing is caring!

With the potential SQL views as yet another schema object type, code needs to be duplicated - or better factored into shared class templates.

#### In this PR:
* Restructured classes representing table database objects:
  * Factored out common functionality into base classes and mixins.
  * Split CTE table class into its own structure, effectively decoupling it from the base table DBO.
  * Moved the facilities for virtual tables into a dedicated header file.
  * Recomposition of the facilities for virtual tables:
    * Split virtual table facilities into: a module tag, virtual table traits that can be specialized, an intermediary description resulting from a `using module` expression and a concrete virtual table class.
* Partially renamed structures by removing the `_t` suffix. Since the advent of alias template for type aliases IMO it isn't the best choice to use it for the library's regular types.
  * `virtual_table_t` -> `virtual_table`.
  * Renamed `table_t` to `base_table` denoting that it is a table that actually stores data. "base table" is the best word in SQL land for this.
  * The renaming should be done for all other DBOs (`index_t`, `trigger_t`) if we agree on this different naming convention. Let me know what you think [yes/no/revert].

#### Additionally:
* Ability to create the FTS5 virtual table using table references.

#### Further changes since the last review on Sep 30 2025:
* Separated generic facilities for virtual tables from facilities for concrete virtual table provided by modules/extensions.
* Explicit object mapping for a virtual table must now be specified when calling `make_virtual_table<O>()` instead of when calling `using_module<O>()`.
* Converted the `dbstat` table into an eponymous virtual table, additionally with the option of creating further instances of it.

trueqbit committed:
Restructured classes representing table database objects

* Factored out common functionality into common base classes and mixins.
* Renamed structures.
* Split CTE table class into its own structure, effectively decoupling it from a table DBO.

trueqbit committed:
Split virtual table into a dedicated header file

trueqbit committed:
Ability to create the FTS5 virtual table extension using table references

trueqbit committed:
Merge branch 'upstream/dev' into upstream/restructure-table-dbos

trueqbit committed:
Split regular and virtual table constraint trait check

trueqbit committed:
Renamed `mapped_columns_mixin` -> `table_definition`

trueqbit committed:
Recomposition of the facilities for virtual tables

* Split virtual table facilities into: a module tag, virtual table traits that can be specialized, an intermediary description resulting from a `using module` expression and a concrete virtual table class.
* Renamed table DBO classes.
* Removed the `_t` suffix from table DBO classes and renamed `virtual_table_t` -> `virtual_table` and `table_t` -> `base_table`, where the latter denotes a table that actually stores data.
* Finally replaced the 'without rowid' NTTP by a type parameter.

trueqbit committed:
Used `std::true_type`/`std::false_type` instead of `polyfill::bool_constant`

trueqbit committed:
Corrected unit test

fnc12 commented on dev/schema/table_base.h:
```diff
@@ -0,0 +1,246 @@
+    struct mapped_object_mixin {
+        using object_type = O;
+
+        /**
```
this formatting is sad =(

trueqbit replied:
Hmm, what exactly do you mean? Do you want me to change something?

fnc12 replied:
I am talking about different indentations on the first two lines. I don't want you to change something until formatter CI tells everything is ok. It is not the first time formatter produces weird output. I think it is something we can't avoid. But of you know how to fix it fast it would be great. If you merge it right now without changing - ok as well.

fnc12 reviewed (approved):
The naming is ok. The PR is good. Thank you @trueqbit

trueqbit committed:
Explicitly mapped object must be specified when making the virtual table

... instead of when making the definition of the table using a specific module.
Additionally:
* Split virtual table modules into separate header files.

trueqbit committed:
Moved basic virtual table traits into a base class

trueqbit committed:
Laid the groundwork for instances of eponymous virtual tables

trueqbit committed:
Converted the dbstat table into an eponymous virtual table

trueqbit committed:
Renamed dbstat test cases to avoid duplicates

trueqbit committed:
Updated CPP guard for FTS5 virtual table

fnc12 commented on dev/schema/table_base.h:
```diff
@@ -0,0 +1,246 @@
+         *  @param lambda Lambda called for each column. Function signature: `void(auto& column)`
+         */
+        template<class L>
+        void for_each_column(L&& lambda) const {
```
why do we pass an r-value here? if we don't call `std::move(lambda)` on the next line then it goes further as an l-value. In that case the smarter way is just passing an l-value, probably const l-value

trueqbit replied:
Well, it's just taking anything, which is usually a lambda in its real sense, i.e. created ad-hoc. In the library's logic there's nNo need to pass it on as an rvalue, so we just pass it on as an lvalue. const-qualifying it is not good as the lambda might be modifyable.

fnc12 commented on dev/schema/virtual_table.h:
```diff
@@ -0,0 +1,151 @@
+     *  
+     *  [Deprecation notice] This factory function is deprecated and will be removed in v1.11.
+     */
+    template<class O, class M, class... Cs>
```
can we add deprecate C++ tag here to produce compiler's warnings?

trueqbit replied:
I already added the deprecation warning to the `using_fts5<O>()` factory function. This should be enough, otherwise you get two warnings...

fnc12 commented on dev/storage.h:
```diff
@@ -1188,44 +1187,45 @@ namespace sqlite_orm {
 
                 const auto res = sync_schema_result::already_in_sync;
-                context_t context{this->db_objects};
+                const serializer_context<db_objects_type> context{this->db_objects};
```
like for `const`ness here
